crypto/cmp/cmp_vfy.c

```diff
@@ -0,0 +1,868 @@
+        err = 2;
+        goto cert_err;
+    }
+
+    pubkey = X509_get_pubkey(cert);
+    if (pubkey == NULL) {
+        CMPerr(0, CMP_R_FAILED_EXTRACTING_PUBKEY);
+        err = 2;
+        goto cert_err;
+    }
+
+    /* create the DER representation of protected part */
+    prot_part.header = msg->header;
+    prot_part.body = msg->body;
+
+    l = i2d_CMP_PROTECTEDPART(&prot_part, &prot_part_der);
+    if (l < 0 || prot_part_der == NULL)
+        goto end;
+    prot_part_der_len = (size_t) l;
+
+    /* verify signature of protected part */
+    if (!OBJ_find_sigid_algs(OBJ_obj2nid(msg->header->protectionAlg->algorithm),
+                             &digest_nid, &pk_nid)
+            || digest_nid == NID_undef || pk_nid == NID_undef
+            || (digest = (EVP_MD *)EVP_get_digestbynid(digest_nid)) == NULL) {
+        CMPerr(0, CMP_R_ALGORITHM_NOT_SUPPORTED);
+        err = 2;
+        goto end;
+    }
+
+    /* check msg->header->protectionAlg is consistent with public key type */
+    if (EVP_PKEY_type(pk_nid) != EVP_PKEY_base_id(pubkey)) {
+        CMPerr(0, CMP_R_WRONG_ALGORITHM_OID);
+        err = 2;
+        goto end;
+    }
+
+    if ((ctx = EVP_MD_CTX_new()) == NULL)
+        goto end;
+    err = (EVP_VerifyInit_ex(ctx, digest, NULL)
+               && EVP_VerifyUpdate(ctx, prot_part_der, prot_part_der_len)
+               && EVP_VerifyFinal(ctx, msg->protection->data,
+                                  msg->protection->length, pubkey) == 1)
+        ? 0 : 2;
+
+ end:
+    EVP_MD_CTX_free(ctx);
+    OPENSSL_free(prot_part_der);
+    EVP_PKEY_free(pubkey);
+
+ cert_err:
+    if (err == 2) { /* print diagnostics on cert verification error */
+        X509_STORE *ts = cmp_ctx->trusted;
+        X509_STORE_CTX *csc = X509_STORE_CTX_new();
+        X509_STORE_CTX_verify_cb verify_cb = ts == NULL ? NULL
+                                             : X509_STORE_get_verify_cb(ts);
+
+        if (csc != NULL && verify_cb != NULL /* implies ts != NULL */
+                && X509_STORE_CTX_init(csc, ts, NULL, NULL)) {
+            X509_STORE_CTX_set_current_cert(csc, cert);
+            X509_STORE_CTX_set_error_depth(csc, -1);
+            X509_STORE_CTX_set_error(csc, X509_V_ERR_UNSPECIFIED);
+            (void)(*verify_cb)(0, csc);
+        }
+        CMPerr(0, CMP_R_ERROR_VALIDATING_PROTECTION);
+        X509_STORE_CTX_free(csc);
+    }
+    return err == 0;
+}
+
+/* Verify a message protected with PBMAC */
+static int CMP_verify_PBMAC(const OSSL_CMP_MSG *msg,
+                            const ASN1_OCTET_STRING *secret)
+{
+    ASN1_BIT_STRING *protection = NULL;
+    int valid = 0;
+
+    /* generate expected protection for the message */
+    if ((protection = ossl_cmp_calc_protection(msg, secret, NULL)) == NULL)
+        return 0;               /* failed to generate protection string! */
+
+    valid = ASN1_STRING_cmp((const ASN1_STRING *)protection,
+                            (const ASN1_STRING *)msg->protection) == 0;
```

Comment:
Should we be doing a CRYPTO_memcmp here?

Reply:
Here we compare ASN1_BIT_STRING's, should be OK?

Reply:
The question really is whether this comparison should be happening in constant time or not (ASN1_STRING_cmp is not constant time, but CRYPTO_memcmp is).

Reply:
I changed it to CRYPTO_memcmp(), fixed